test/adaptive_load/plugin_util_test.cc

```diff
@@ -0,0 +1,378 @@
+ *
+ * @param message An Any proto that must wrap a LinearScoringFunctionConfig.
+ *
+ * @return Status InvalidArgument if threshold is kBadConfigThreshold, OK otherwise.
+ */
+absl::Status DoValidateConfig(const Envoy::Protobuf::Message& message) {
+  const auto& any = dynamic_cast<const Envoy::ProtobufWkt::Any&>(message);
+  nighthawk::adaptive_load::LinearScoringFunctionConfig config;
+  Envoy::MessageUtil::unpackTo(any, config);
+  return config.threshold() == kBadConfigThreshold
+             ? absl::InvalidArgumentError("input validation failed")
+             : absl::OkStatus();
+}
+
+/**
+ * InputVariableSetter for testing.
+ */
+class TestInputVariableSetter : public InputVariableSetter {
+public:
+  // Any plugin in the adaptive load system can freely choose an arbitrary single proto as its
+  // config type. We use LinearScoringFunctionConfig for all plugins in this test.
+  TestInputVariableSetter(const nighthawk::adaptive_load::LinearScoringFunctionConfig& config)
+      : value_from_config_proto_{config.threshold()} {}
+  absl::Status SetInputVariable(nighthawk::client::CommandLineOptions& command_line_options,
+                                double input_value) override {
+    command_line_options.mutable_connections()->set_value(static_cast<unsigned int>(input_value));
+    return absl::OkStatus();
+  }
+  const double value_from_config_proto_;
+};
+
+/**
+ * A factory that creates a TestInputVariableSetter from a LinearScoringFunctionConfig (see
+ * TestInputVariableSetter constructor).
+ */
+class TestInputVariableSetterConfigFactory : public InputVariableSetterConfigFactory {
+public:
+  std::string name() const override { return "nighthawk.test-input-variable-setter"; }
+  absl::Status ValidateConfig(const Envoy::Protobuf::Message& message) const override {
+    return DoValidateConfig(message);
+  }
+
+  Envoy::ProtobufTypes::MessagePtr createEmptyConfigProto() override {
+    return std::make_unique<nighthawk::adaptive_load::LinearScoringFunctionConfig>();
+  }
+
+  InputVariableSetterPtr
+  createInputVariableSetter(const Envoy::Protobuf::Message& message) override {
+    const auto& any = dynamic_cast<const Envoy::ProtobufWkt::Any&>(message);
+    nighthawk::adaptive_load::LinearScoringFunctionConfig config;
+    Envoy::MessageUtil::unpackTo(any, config);
+    return std::make_unique<TestInputVariableSetter>(config);
+  }
+};
+
+REGISTER_FACTORY(TestInputVariableSetterConfigFactory, InputVariableSetterConfigFactory);
+
+/**
+ * ScoringFunction for testing.
+ */
+class TestScoringFunction : public ScoringFunction {
+public:
+  // Any plugin in the adaptive load system can freely choose an arbitrary single proto as its
+  // config type. We use LinearScoringFunctionConfig for all plugins in this test.
+  TestScoringFunction(const nighthawk::adaptive_load::LinearScoringFunctionConfig& config)
+      : value_from_config_proto_{config.threshold()} {}
+  double EvaluateMetric(double) const override { return 1.0; }
```

Comment:
(nit, optional) Can we add newlines between the constructor, the individual methods and the members so we have three distinct groups? It may improve readability.

If we choose to do it here, let's apply it to all the relevant objects.

Reply:
Done